File: apps/dsa-web/src/api/chipHealth.ts
import apiClient from './index';

export interface HoldingItem {
  code: string;
  name?: string;
  cost?: number | null;
}

export interface ChipHealthRow {
  code: string;
  name: string;
  data_date: string;
  close: number | null;
  cost: number | null;
  pnl: number | null;
  ret5: number | null;
  level: string; // 🔴/🟠/🟡/⚪/🟢/—
  verdict: string;
  profit: number | null;
  bottom: number | null;
  loss3: number | null;
  loss5: number | null;
  loss7: number | null;
  high_profit_risk: boolean;
}

export interface ChipHealthResult {
  generated_at: string;
  data_cutoff: string;
  count: number;
  results: ChipHealthRow[];
  skipped: { code: string; name: string; reason: string }[];
}

export interface ChipHealthStatus {
  data_dir: string;
  cache_exists: boolean;
  cache_mtime: string | null;
  holdings_count: number;
  holdings_file: string;
}

export const chipHealthApi = {
  async status(): Promise<ChipHealthStatus> {
    const response = await apiClient.get('/api/v1/chip-health/status');
    return response.data;
  },

  async getHoldings(): Promise<HoldingItem[]> {
    const response = await apiClient.get('/api/v1/chip-health/holdings');
    return response.data.holdings || [];
  },

  async saveHoldings(holdings: HoldingItem[]): Promise<number> {
    const response = await apiClient.put('/api/v1/chip-health/holdings', { holdings });
    return response.data.count;
  },

  async run(holdings?: HoldingItem[]): Promise<ChipHealthResult> {
    // 引擎需载入1GB缓存, 单独放宽超时
    const response = await apiClient.post(
      '/api/v1/chip-health/run',
      holdings ? { holdings } : {},
      { timeout: 300000 },
    );
    return response.data;
  },
};

File: api/v1/schemas/chip_health.py
# -*- coding: utf-8 -*-
"""筹码体检接口模型"""

from typing import Any, Dict, List, Optional

from pydantic import BaseModel, Field


class HoldingItem(BaseModel):
    code: str = Field(..., description="6位股票代码", example="600519")
    name: Optional[str] = Field("", description="股票名称(可空,自动补全)")
    cost: Optional[float] = Field(None, description="成本价(可空,填了计算浮盈)")


class HoldingsSaveRequest(BaseModel):
    holdings: List[HoldingItem] = Field(default_factory=list, description="持仓清单(全量覆盖保存)")


class HoldingsListResponse(BaseModel):
    holdings: List[HoldingItem]
    count: int


class ChipHealthRunRequest(BaseModel):
    holdings: Optional[List[HoldingItem]] = Field(
        None, description="临时体检清单(不传则用已保存的持仓清单)")


class ChipHealthRunResponse(BaseModel):
    generated_at: str
    data_cutoff: str = Field("", description="缓存数据截止日")
    results: List[Dict[str, Any]]
    skipped: List[Dict[str, Any]] = Field(default_factory=list)
    count: int = 0


class ChipHealthStatusResponse(BaseModel):
    data_dir: str
    cache_exists: bool
    cache_mtime: Optional[str] = None
    holdings_count: int = 0
    holdings_file: str = ""
